tests/failpoints/cases/test_transfer_leader.rs

```diff
@@ -680,3 +682,286 @@ fn test_check_long_uncommitted_proposals_after_became_leader() {
+    let _ = cluster.async_put(b"k2", b"v2").unwrap();
+
+    pd_client.transfer_leader(region_id, new_peer(2, 2), vec![]);
+    sleep_ms(300);
+    assert_eq!(
+        pd_client.check_region_leader(region_id, new_peer(2, 2)),
+        false
+    );
+
+    pd_client.transfer_leader(region_id, new_peer(3, 3), vec![]);
+    sleep_ms(300);
+    assert_eq!(
+        pd_client.check_region_leader(region_id, new_peer(3, 3)),
+        false
+    );
+
+    pd_client.transfer_leader(region_id, new_peer(4, 4), vec![]);
+    pd_client.region_leader_must_be(region_id, new_peer(4, 4));
+
+    // Verify the data completeness on the new leader.
+    must_get_equal(&cluster.get_engine(4), b"k1", b"v1");
+    must_get_equal(&cluster.get_engine(4), b"k2", b"v2");
+
+    pd_client.must_none_peer(region_id, new_peer(2, 2));
+}
+
+// This test verifies that a leader transfer is rejected when the transferee
+// has been demoted to a learner but the leader has not yet applied this
+// configuration change.
+#[test]
+fn test_when_applied_conf_change_on_learner_transferee() {
+    let mut cluster = new_server_cluster(0, 3);
+    let pd_client = cluster.pd_client.clone();
+    pd_client.disable_default_operator();
+    let region_id = cluster.run_conf_change();
+    pd_client.must_add_peer(region_id, new_peer(2, 2));
+    pd_client.must_add_peer(region_id, new_peer(3, 3));
+    pd_client.region_leader_must_be(region_id, new_peer(1, 1));
+
+    fail::cfg("apply_on_conf_change_1_1", "pause").unwrap();
+
+    // Demote peer-2 to be a learner.
+    pd_client.joint_confchange(
+        region_id,
+        vec![(ConfChangeType::AddLearnerNode, new_learner_peer(2, 2))],
+    );
+    sleep_ms(300);
+
+    pd_client.transfer_leader(region_id, new_peer(2, 2), vec![]);
+    sleep_ms(300);
+    assert_eq!(
+        pd_client.check_region_leader(region_id, new_peer(2, 2)),
+        false
+    );
+
+    pd_client.transfer_leader(region_id, new_peer(3, 3), vec![]);
+    pd_client.region_leader_must_be(region_id, new_peer(3, 3));
+    let region = block_on(pd_client.get_region_by_id(region_id))
+        .unwrap()
+        .unwrap();
+    assert_eq!(region.get_peers()[1].get_role(), PeerRole::Learner);
+}
+
+// This test verifies that a leader transfer is allowed when the transferee
+// has been applied a conf change but the leader has not yet applied.
+#[test]
+fn test_when_applied_conf_change_on_transferee_pessimistic_lock() {
+    let mut cluster = new_server_cluster(0, 4);
+    let pd_client = cluster.pd_client.clone();
+    pd_client.disable_default_operator();
+    let region_id = cluster.run_conf_change();
+    pd_client.must_add_peer(region_id, new_peer(2, 2));
+    pd_client.must_add_peer(region_id, new_peer(3, 3));
+    pd_client.region_leader_must_be(region_id, new_peer(1, 1));
+
+    fail::cfg("apply_on_conf_change_1_1", "pause").unwrap();
+    fail::cfg("propose_locks_before_transfer_leader", "return").unwrap();
```

Comment:
Just curious, why is this failpoint necessary? What happens if we don't set it? Would be great to have a line of comment about it.

Reply:
This failpoint is used to simulate the transfer-leader with pessimistic locks scenario, enabling us to propose a TransferLeader command. 
IMO maybe we don't need a line comment since it is not particularly complex? In addition, this failpoint is used in multiple place, I don't have an ideal way to add a comment for it :)

Reply:
Yes. But for 'twice', to be more precise, the first time refers to sending the MsgTransferLeader Raft message, and the second time refers to proposing the TransferLeader Raft command.”